### PR TITLE
Add nested ordering filter

### PR DESCRIPTION
**Pull request by xiangge**
Add nested ordering filter to instead original one
in rest framework.
Update doc of ordering part.

JIRA: PDC-1746

### CODE REVIEW

File: pdc/apps/common/viewsets.py [hunk 1]
```diff
@@ -205,7 +205,7 @@ def _check_ordering_keys(self, request):
         serializer_fields = self._get_fields_from_serializer_class()
         valid_fields = list(set(model_fields).intersection(set(serializer_fields)))
         valid_fields += self.queryset.query.aggregates.keys()
-        tmp_list = [param.strip().lstrip('-') for param in ordering_keys.split(',')]
+        tmp_list = [param.strip().lstrip('-').split("__")[0] for param in ordering_keys.split(',')]
```
Comment (hluk) [inline pdc/apps/common/viewsets.py]: I think this shouldn't be here. Instead the available filter fields should be extended (e.g. `ReleaseComponentRelationshipFilter.Meta.fields`).

File: pdc/apps/utils/utils.py [hunk 2]
```diff
@@ -59,3 +61,33 @@ def read_permission_for_all():
 
 def get_model_name_from_obj_or_cls(obj_or_cls):
     return ContentType.objects.get_for_model(obj_or_cls).model
+
+
+class RelatedNestedOrderingFilter(OrderingFilter):
+    """
+    Extends OrderingFilter to support ordering by fields in related models
+    using Django orm '__', this change is based on rest_framework 3.2.5
```
Comment (hluk) [inline pdc/apps/utils/utils.py]: If it works in rest_framework 3.2.5, why is this class needed? I don't see where the actual ordering is done. Shouldn't `filter_queryset()` method be also implemented?

**Comment by xiangge** [comment]
@hluk For the test part, I have thought about it, because it is a common class, finally all the rest apis will  use this class, that means if all the rest apis' test cases work, this class is okay and the test is passed.

**Comment by xiangge** [comment]
@hluk I will add an test case to test the nested ordering here to check if this class work.
Like <server_url>/?orderging=to_component__name.

File: pdc/apps/utils/utils.py (hunk 2 above)
Comment (xiangge) [inline pdc/apps/utils/utils.py]: In 3.2.5, the OrderingFilter just works for the fields in model, but it can't ordering by the nested fields. 
Like there are four values in model(ReleaseComponentRelationship) can be ordered: 
['relation_type', 'id', 'from_component', 'to_component']
from_component and to_component are foreign keys.
The values here:
"to_component": {
        "id": "int", 
        "name": "string", 
        "release": "Release.release_id"
    }, 
And as default if you ordered by from_component or to_component, you can just ordered by the id of them. And you can just call <server_url>/?orderging=to_component.

I wrote this class to let people can use such apis, like <server_url>/?orderging=to_component__name. or <server_url>/?orderging=to_component__release.

File: pdc/apps/common/viewsets.py (hunk 1 above)
Comment (xiangge) [inline pdc/apps/common/viewsets.py]: As I mentioned here, if there is no this changes, if I call <server_url>/?orderging=to_component__name.
 It will raise error like:

Request Method: | GET
-- | --
http://10.66.128.226:8080/rest_api/v1/release-component-relationships/?ordering=-to_component__name
1.8.18
FieldError
Unknown query key: [u'to_component__name'] not in fields: ['relation_type', 'id', 'from_component', 'to_component']

I think this function is just check if the fields in model are correct. And if it includes nested files with "__", will check it in the funciton  is_valid_field of RelatedNestedOrderingFilter.

So after this change:
if you use a url like:
<server_url>/?orderging=to_componentsadfsa
It will raise errors:

Request Method: | GET
-- | --
http://10.66.128.226:8080/rest_api/v1/release-component-relationships/?ordering=-to_componentasdf

And if you use:
<server_url>/?orderging=to_component__namesadfsa.
will raise error like:

Request Method: | GET
-- | --
http://10.66.128.226:8080/rest_api/v1/release-component-relationships/?ordering=-to_component__nameasdfads
1.8.18

Comment (hluk) [inline pdc/apps/common/viewsets.py]: Hmm, following should give me an error but doesn't.

```
pdc_client -s local -r release-component-relationships -d '{"ordering":"from_component__zzz"}'
```

Comment (xiangge) [inline pdc/apps/common/viewsets.py]: Hmmm, in my local env, it can raise error, a little strange：
 pdc_client -s http://<server>:8080/rest_api/v1 -r release-component-relationships -d '{"ordering":"from_component__zzz"}'

500 b'\n<!DOCTYPE html>\n<html lang="en">\n<head>\n  <meta http-equiv="content-type" content="text/html; charset=utf-8">\n  <meta name="robots" content="NONE,NOARCHIVE">\n  <title>FieldError at /rest_api/v1/release-component-relationships/</title>\n  <style type="text/css">\n    html * { padding:0; margin:0; }\n    body * { padding:10px 20px; }\n    body * * { padding:0; }\n    body { font:small sans-serif; }\n    body>div { border-bottom:1px solid #ddd; }\n    h1 { font-weight:normal; }\n    h2 { margin-bottom:.8em; }\n   ........

 pdc_client -s http://<server>:8080/rest_api/v1 -r release-component-relationships -d '{"ordering":"from_component__name"}'
{
    "count": 3,
    "next": null,
    "previous": null,
    "results": [
        {
            "from_component": {
                "id": XXX,
                "name": "XXX",
                "release": "XXXX"
            },
......

**Comment by bliuredhat** [comment]
Looks good for me; and adding one test case for this patch will be more better.

File: pdc/settings.py [hunk 3]
```diff
@@ -93,7 +93,7 @@
     'DEFAULT_METADATA_CLASS': 'contrib.bulk_operations.metadata.BulkMetadata',
 
     'DEFAULT_FILTER_BACKENDS': ('rest_framework.filters.DjangoFilterBackend',
-                                'rest_framework.filters.OrderingFilter'),
+                                'pdc.apps.utils.utils.RelatedNestedOrderingFilter'),
```
Comment (bliuredhat) [inline pdc/settings.py]: Just a reminder; you need to update the settings_local.py.dist  and make a sync with the DEFAULT_FILTER_BACKENDS value.

Comment (simozhan) [inline pdc/settings.py]: Yeah. @bliuredhat is right. we need to update the local dist file as well, so that the user can use local.py correctly

File: pdc/apps/common/viewsets.py (hunk 1 above)
Comment (simozhan) [inline pdc/apps/common/viewsets.py]: I am wondering if we can skip nested ordering check here and let is_valid_field do the check?
I am not sure

File: pdc/settings.py (hunk 3 above)
Comment (chcao55) [inline pdc/settings.py]: OK, Good idea @bliuredhat and @simozhan , will do another commit with the changes

File: pdc/apps/common/renderers.py [hunk 4]
```diff
@@ -66,6 +66,7 @@
 
 URL_SPEC_RE = re.compile(r'\$(?P<type>URL|LINK):(?P<details>[^$]+)\$')
 ORDERING_STRING = "\n * `ordering` is used to override the ordering of the results, the value could be: "
+ORDERING_STRING_NESTED = "The nested ordering is supported by using \_\_ ."
```
Comment (chcao55) [inline pdc/apps/common/renderers.py]: @simozhan  Good idea, will do another commit with the changes.

File: pdc/apps/common/viewsets.py (hunk 1 above)
Comment (xiangge) [inline pdc/apps/common/viewsets.py]: @simozhan good idea, I will check it in next commit.

**Comment by chcao55** [comment]
@hluk @simozhan @bliuredhat , please help to review again, all changes are added,

File: pdc/apps/component/tests.py [hunk 5]
```diff
@@ -1866,6 +1866,25 @@ def test_delete_group(self):
         response = self.client.get(url, format='json')
         self.assertEqual(response.status_code, status.HTTP_404_NOT_FOUND)
 
+    def test_list_relationship_by_nested(self):
```
Comment (simozhan) [inline pdc/apps/component/tests.py]: Because this feature will affects all apis ordering. So is it ok to add some exceptional test cases like  non-existing-field__name or something

Comment (xiangge) [inline pdc/apps/component/tests.py]: Yes, have added the not existing filed check in test case, please review it again, thanks.

File: pdc/apps/component/tests.py [hunk 6]
```diff
@@ -1866,6 +1866,32 @@ def test_delete_group(self):
...
+            else:
+                self.assertTrue(self, data[0].get('to_component')["release"] >= data[1].get('to_component')["release"])
+
+        for v in ["to_component__name", "-to_component__name", "to_component__release", "-to_component__release"]:
+            f(v)
+
+    def test_list_relationship_non_exist_nested(self):
+        for v in ["to_component_non_exist__release", "to_component__release_non_exist"]:
+            response = self.client.get(reverse("rcrelationship-list"), {'ordering': v}, format='json')
+            if v in ["to_component_non_exist__release", "to_component__release_non_exist"]:
```
Comment (hluk) [inline pdc/apps/component/tests.py]: This line doesn't seem to be required anymore.

File: pdc/apps/utils/utils.py [hunk 7]
```diff
@@ -59,3 +61,43 @@ def read_permission_for_all():
...
     return ContentType.objects.get_for_model(obj_or_cls).model
+
+
+class RelatedNestedOrderingFilter(OrderingFilter):
+    """
+    Extends OrderingFilter to support ordering by fields in related models
+    using Django orm '__'.
+
+    This class is copied and changed from the patch of the github issue.
+    https://github.com/encode/django-rest-framework/issues/1005
```
Comment (hluk) [inline pdc/apps/utils/utils.py]: OK, now I get it. Can you use the updated code from [the latest comment](https://github.com/encode/django-rest-framework/issues/1005#issuecomment-289555282) and mention the comment link in the docstring instead?

File: pdc/apps/component/tests.py [hunk 8]
```diff
@@ -1866,6 +1866,32 @@ def test_delete_group(self):
...
+                self.assertTrue(self, data[0].get('to_component')["name"] <= data[1].get('to_component')["name"])
+            elif ordering_value == "-to_component__name":
+                self.assertTrue(self, data[0].get('to_component')["name"] >= data[1].get('to_component')["name"])
+            elif ordering_value == "to_component__release":
+                self.assertTrue(self, data[0].get('to_component')["release"] <= data[1].get('to_component')["release"])
+            else:
+                self.assertTrue(self, data[0].get('to_component')["release"] >= data[1].get('to_component')["release"])
+
+        for v in ["to_component__name", "-to_component__name", "to_component__release", "-to_component__release"]:
+            f(v)
```
Comment (hluk) [inline pdc/apps/component/tests.py]: Can you test ordering by multiple fields?

File: pdc/apps/component/tests.py (hunk 6 above)
Comment (xiangge) [inline pdc/apps/component/tests.py]: Yes, will remove it.

File: pdc/apps/utils/utils.py (hunk 7 above)
Comment (hluk) [inline pdc/apps/utils/utils.py]: Well the latest comment takes into account "reverse relation". But ignore it if it won't work for with the version of rest-framework we use.

Since there is lot of comments with code in the issue referenced by the docstring, a link to exact comment with the code your change was based on would help a bit more.

Comment (xiangge) [inline pdc/apps/utils/utils.py]: Just check, we don't need add the "reverse relation", the instance is not 'ForeignObjectRel' or 'OneToOneRel'.

And I have already changed the url to the comment which includes the patch,  good suggestion!

File: pdc/apps/component/tests.py (hunk 8 above)
Comment (hluk) [inline pdc/apps/component/tests.py]: @xiangge What I mean is sorting with e.g. `to_component__name,to_component__release` (might not be necessary though).

Comment (xiangge) [inline pdc/apps/component/tests.py]: @hluk  I added a new test function here, called test_list_relationship_multi_nested, which ordered by  to_component__release,to_component__name and -to_component__name,to_component__release.
Do you mean that?

**Comment by simozhan** [comment]
:+1:

**Comment by xiangge** [comment]
Close this pr as patches merged.
